app/scripts/modules/kubernetes/src/common/tolerations/TolerationsComponent.tsx

```diff
@@ -0,0 +1,210 @@
+import * as React from 'react';
+import { cloneDeep } from 'lodash';
+import Select, { Option } from 'react-select';
+import { Button } from 'react-bootstrap';
+import { BindAll } from 'lodash-decorators';
+import { HelpField } from '@spinnaker/core';
+
+export type IOperator = 'Exists' | 'Equal';
+export type IEffect = 'NoSchedule' | 'PreferNoSchedule' | 'NoExecute';
+
+export interface IToleration {
+  key: string;
+  operator: IOperator;
+  effect: IEffect;
+  value: string;
+  tolerationSeconds?: number;
+}
+
+export interface IKubernetesTolerationsProps {
+  tolerations: IToleration[];
+  onTolerationChange: (tolerations: IToleration[]) => void;
+}
+
+export interface IKubernetesTolerationsState {
+  tolerations: IToleration[];
+}
+
+@BindAll()
+export class KubernetesTolerations extends React.Component<IKubernetesTolerationsProps, IKubernetesTolerationsState> {
+  private static operators: Option[] = [{ value: 'Exists', label: 'Exists' }, { value: 'Equal', label: 'Equal' }];
+
+  private static effects: Option[] = [
+    { value: 'NoSchedule', label: 'NoSchedule' },
+    { value: 'PreferNoSchedule', label: 'PreferNoSchedule' },
+    { value: 'NoExecute', label: 'NoExecute' },
+  ];
+
+  public tolerations: IToleration[];
+
+  constructor(props: IKubernetesTolerationsProps) {
+    super(props);
+    this.state = this.getState(props);
+  }
+
+  private getState(props: IKubernetesTolerationsProps): IKubernetesTolerationsState {
+    return {
+      tolerations: props.tolerations || [],
+    };
+  }
+
+  public componentWillReceiveProps(nextProps: IKubernetesTolerationsProps): void {
+    this.setState(this.getState(nextProps));
+  }
+
+  private addTolerations(): void {
+    const tolerations = cloneDeep(this.state.tolerations);
+    tolerations.push({ key: '', value: '', operator: 'Exists', effect: 'NoSchedule' });
+    this.props.onTolerationChange(tolerations);
+  }
+
+  private removeToleration(index: number): () => void {
+    return () => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations.splice(index, 1);
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  private handleKeyChange(index: number): (event: any) => void {
```

Comment:
These methods can be replaced by a single method in a TS-friendly way:

```
private handleChange(index: number, tolerationProperty: keyof IToleration): (event: any) => void {
  return (event: any) => {
    const tolerations = cloneDeep(this.state.tolerations);
    tolerations[index][tolerationProperty] = event.target.value;
    this.props.onTolerationChange(tolerations);
  };
}
```

And called below as `this.handleChange(i, 'effect')`.

Reply:
That's one of the changes i thought about making. Just wanted to get some feedback first! Agreed!